### PR TITLE
homework2 ver. 1.0

### PR DESCRIPTION
**Pull request by neme3is**

### CODE REVIEW

File: src/test/java/com/epam/tc/hw2/exercise1/ExerciseOneTests.java [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+        // Assert that there are 4 items on the header section are displayed and they have proper texts
+        softAssert.assertEquals(webDriver.findElements(By
+            .xpath("//ul[@class='uui-navigation nav navbar-nav m-l8']/li/a")).size(), 4);
+        webDriver.findElements(By.xpath("//ul[@class='uui-navigation nav navbar-nav m-l8']/li/a"))
+                 .forEach(i -> softAssert.assertTrue(i.isDisplayed()));
+        softAssert.assertEquals(webDriver.findElements(By
+            .xpath("//ul[@class='uui-navigation nav navbar-nav m-l8']/*[contains(text(),'HOME') "
+            + "or contains(text(),'CONTACT FORM') or contains(text(),'SERVICE') "
+                + "or contains(text(),'METALS & COLORS')]")).size(), 4);
```
Comment (shoggy) [inline src/test/java/com/epam/tc/hw2/exercise1/ExerciseOneTests.java:34]: Please choose more suitable locators. 
Remember order: DOM, CSS, XPath

File: src/test/java/com/epam/tc/hw2/exercise1/ExerciseOneTests.java [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+        // Assert that there are 4 images on the Index Page and they are displayed
+        webDriver.findElements(By.cssSelector("benefit-icon")).forEach(i -> softAssert.assertTrue(i.isDisplayed()));
```
Comment (shoggy) [inline src/test/java/com/epam/tc/hw2/exercise1/ExerciseOneTests.java:37]: Using loops inside for validation is a bad practice (in most cases). You can use AssertJ that has a matcher for collections (something like `assertThat(Collection).allMatch(...)`

File: src/test/java/com/epam/tc/hw2/exercise1/ExerciseOneTests.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+        // Assert that there are 4 texts on the Index Page under icons and they have proper text
+        ArrayList<String> texts = new ArrayList<>(Arrays
```
Comment (shoggy) [inline src/test/java/com/epam/tc/hw2/exercise1/ExerciseOneTests.java:40]: Please use interface for variable declaration and Java 9 has a static method `of(...)` that makes code better looking

File: src/test/java/com/epam/tc/hw2/exercise1/ExerciseOneTests.java [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+            .asList("To include good practices and ideas from successful EPAM project",
+            "To be flexible and customizable", "To be multiplatform",
+                "Already have good base (about 20 internal and some external projects), wish to get more…"));
+        List<WebElement> list = webDriver.findElements(By.className("benefit-txt"));
+        softAssert.assertEquals(list.get(0).getAttribute("innerHTML").split("<br>"), texts.get(0));
+        softAssert.assertEquals(list.get(1).getAttribute("innerHTML").split("<br>"), texts.get(1));
+        softAssert.assertEquals(list.get(2).getAttribute("innerHTML").split("<br>"), texts.get(2));
+        softAssert.assertEquals(list.get(3).getAttribute("innerHTML").split("<br>"), texts.get(3));
+        list.forEach(i -> softAssert.assertTrue(i.isDisplayed()));
```
Comment (shoggy) [inline src/test/java/com/epam/tc/hw2/exercise1/ExerciseOneTests.java:49]: Please rewrite it using loop or streams and use matcher for collection

File: src/test/java/com/epam/tc/hw2/exercise1/ExerciseOneTests.java [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+        // Assert that there is the iframe with “Frame Button” exist
+        softAssert.assertTrue(webDriver.findElements(By
+            .xpath("//iframe[starts-with(@src, 'https://jdi-testing.github.io/jdi-light/frame-button.html')]"))
+                                       .size() > 0);
```
Comment (shoggy) [inline src/test/java/com/epam/tc/hw2/exercise1/ExerciseOneTests.java:54]: I think there is more suitable locator

File: src/test/java/com/epam/tc/hw2/exercise1/ExerciseOneTests.java [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+
+        // Assert that there are 5 items in the Left Section are displayed and they have proper text
+        List<WebElement> we = webDriver.findElements(By.xpath("//div[@name='navigation-sidebar']//li"));
+        ArrayList<String> texts1 = new ArrayList<>(Arrays.asList("Home", "Contact form", "Service",
+            "Metals & Colors", "Elements packs"));
+        softAssert.assertEquals(we.get(0).getAttribute("innerHTML").split("<br>"), texts1.get(0));
+        softAssert.assertEquals(we.get(1).getAttribute("innerHTML").split("<br>"), texts1.get(1));
+        softAssert.assertEquals(we.get(2).getAttribute("innerHTML").split("<br>"), texts1.get(2));
+        softAssert.assertEquals(we.get(3).getAttribute("innerHTML").split("<br>"), texts1.get(3));
+        softAssert.assertEquals(we.get(4).getAttribute("innerHTML").split("<br>"), texts1.get(4));
```
Comment (shoggy) [inline src/test/java/com/epam/tc/hw2/exercise1/ExerciseOneTests.java:71]: Please use another locator type
Use collections assertion

**Review by shoggy — CHANGES_REQUESTED** [review]
Ex2 has same issues as the ex1, please fix them too

File: pom.xml [hunk 7]
```diff
@@ -38,6 +40,17 @@
             <scope>system</scope>
             <systemPath>${project.basedir}/libs/calculator-1.0.jar</systemPath>
         </dependency>
+        <dependency>
+            <groupId>org.seleniumhq.selenium</groupId>
+            <artifactId>selenium-java</artifactId>
+            <version>${selenium-java.version}</version>
+        </dependency>
```
[the remark this answers is not in the document]
Comment (neme3is) [reply pom.xml:47]: No, we need test scope :)

File: src/test/java/com/epam/tc/hw2/Hw2TestsBaseClass.java [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+        WebDriverManager.chromedriver().setup();
+        webDriver = new ChromeDriver();
+        webDriver.manage().timeouts()
+                 .implicitlyWait(3, TimeUnit.SECONDS);
+        webDriver.manage().window().maximize();
+    }
+
+    @AfterClass(alwaysRun = true)
+    public void afterTestMethod() {
+        webDriver.close();
```
[the remark this answers is not in the document]
Comment (neme3is) [reply src/test/java/com/epam/tc/hw2/Hw2TestsBaseClass.java:24]: Ok, i fixed this.

File: src/test/java/com/epam/tc/hw2/exercise1/ExerciseOneTests.java [hunk 9]
```diff
@@ -1,73 +1,103 @@
...
 import java.util.List;
+import org.assertj.core.api.SoftAssertions;
 import org.openqa.selenium.By;
 import org.openqa.selenium.WebElement;
 import org.testng.annotations.Test;
-import org.testng.asserts.SoftAssert;
 
 public class ExerciseOneTests extends Hw2TestsBaseClass {
 
+    SoftAssertions softAssert = new SoftAssertions();
```
Comment (shoggy) [inline src/test/java/com/epam/tc/hw2/exercise1/ExerciseOneTests.java:12]: Move this variable declaration into Test-method

File: src/test/java/com/epam/tc/hw2/exercise1/ExerciseOneTests.java [hunk 10]
```diff
@@ -1,73 +1,103 @@
...
+                  .isTrue();
+
+        softAssert.assertThat(webDriver.findElements(By.cssSelector(".m-l8 > li > a:first-child")))
+                  .as("Header elements is not displayed")
+                  .allMatch(i -> i.isDisplayed());
+
+        List<String> menuElements = List.of("HOME", "CONTACT FORM", "SERVICE", "METALS & COLORS");
+
+        for (String i : menuElements) {
+            softAssert.assertThat(webDriver.findElement(By.linkText(i))
```
Comment (shoggy) [inline src/test/java/com/epam/tc/hw2/exercise1/ExerciseOneTests.java:44]: I think here we can use css selector that retrieves menu items, not by name

File: src/test/java/com/epam/tc/hw2/exercise1/ExerciseOneTests.java [hunk 11]
```diff
@@ -1,73 +1,103 @@
...
+            "Already have good base (about 20 internal and some external projects), wish to get more…");
         List<WebElement> list = webDriver.findElements(By.className("benefit-txt"));
-        softAssert.assertEquals(list.get(0).getAttribute("innerHTML").split("<br>"), texts.get(0));
-        softAssert.assertEquals(list.get(1).getAttribute("innerHTML").split("<br>"), texts.get(1));
-        softAssert.assertEquals(list.get(2).getAttribute("innerHTML").split("<br>"), texts.get(2));
-        softAssert.assertEquals(list.get(3).getAttribute("innerHTML").split("<br>"), texts.get(3));
-        list.forEach(i -> softAssert.assertTrue(i.isDisplayed()));
+
+        for (int i = 0; i < texts.size(); i++) {
+            softAssert.assertThat(list.get(i).getText().replaceAll("\n", " "))
```
Comment (shoggy) [inline src/test/java/com/epam/tc/hw2/exercise1/ExerciseOneTests.java:65]: Make this conversion of the whole`list` before the assertion and compare two collections (expected and converted list)

**Review by shoggy — CHANGES_REQUESTED** [review]

File: src/test/java/com/epam/tc/hw2/exercise1/ExerciseOneTests.java [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+        webDriver.findElement(By.id("login-button")).click();
+
+        softAssert.assertThat(webDriver.findElement(By.id("user-name")).getText())
+                  .as("Username is not correct")
+                  .isEqualTo("ROMAN IOVLEV");
+
+        // Assert that there are 4 items on the header section are displayed and they have proper texts
+        softAssert.assertThat(webDriver.findElements(By.cssSelector(".m-l8 > li > a:first-child")).size() == 4)
+                  .as("Amount of header elements is not 4")
+                  .isTrue();
```
Comment (shoggy) [inline src/test/java/com/epam/tc/hw2/exercise1/ExerciseOneTests.java:36]: Please use a specific method for collections size verification, something like `hasSize(...)`

File: src/test/java/com/epam/tc/hw2/exercise2/ExerciseTwoTests.java [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+            "Earth: condition changed to true", "Water: condition changed to false");
+
+        List<WebElement> checkboxLogElements = webDriver.findElements(By
+            .cssSelector(".logs li"));
+
+        List<String> logTexts = new ArrayList<>();
+
+        for (int i = 0; i < 4; i++) {
+            logTexts.add(checkboxLogElements.get(i).getText().substring(9));
+        }
```
Comment (shoggy) [inline src/test/java/com/epam/tc/hw2/exercise2/ExerciseTwoTests.java:84]: Please do not use magic numbers (4 and 9 in this case)

**Review by shoggy — CHANGES_REQUESTED** [review]

File: src/test/java/com/epam/tc/hw2/exercise1/ExerciseOneTests.java (hunk 12 above)
Comment (neme3is) [reply src/test/java/com/epam/tc/hw2/exercise1/ExerciseOneTests.java:36]: fixed

File: src/test/java/com/epam/tc/hw2/exercise2/ExerciseTwoTests.java (hunk 13 above)
Comment (neme3is) [reply src/test/java/com/epam/tc/hw2/exercise2/ExerciseTwoTests.java]: fixed

**Review by shoggy — APPROVED** [review]

**Pull request merged**